enterprise/server/sociartifactstore/sociartifactstore.go

```
@@ -174,65 +202,79 @@ func getTargetImageRef(ctx context.Context, image string, platform *rgpb.Platfor
 	if err != nil {
-		return ctrname.Digest{}, err
+		return v1.Hash{}, err
 	}
 
 	// Check whether the supplied credentials are sufficient to access the
 	// remote image.
 	if err := checkAccess(ctx, imageRef.Context(), targetImg, authenticator); err != nil {
-		return ctrname.Digest{}, err
+		return v1.Hash{}, err
 	}
 
-	targetImgDigest, err := targetImg.Digest()
+	manifest, err := targetImg.Manifest()
 	if err != nil {
-		return ctrname.Digest{}, err
+		return v1.Hash{}, err
 	}
-	return imageRef.Context().Digest(targetImgDigest.String()), nil
+	return manifest.Config.Digest, nil
 }
 
 func (s *SociArtifactStore) GetArtifacts(ctx context.Context, req *socipb.GetArtifactsRequest) (*socipb.GetArtifactsResponse, error) {
 	ctx, err := prefix.AttachUserPrefixToContext(ctx, s.env)
 	if err != nil {
 		return nil, err
 	}
-	imageRef, err := getTargetImageRef(ctx, req.Image, req.Platform, req.Credentials)
+	configHash, err := getImageManifestConfig(ctx, req.Image, req.Platform, req.Credentials)
 	if err != nil {
 		return nil, err
 	}
-	imageRefDigest := imageRef.DigestStr()
-	exists, err := s.blobstore.BlobExists(ctx, blobKey(imageRefDigest))
-	if err != nil {
-		return nil, err
-	}
-	if exists {
-		return s.getArtifactsFromCache(ctx, imageRefDigest)
-	}
-	// TODO(iain): add a mutex to prevent multiple parallel calls
-	sociIndexDigest, ztocDigests, err := pullAndIndexImage(ctx, req.Image, imageRef.DigestStr())
-	if err != nil {
-		return nil, err
-	}
-	if err = s.writeArtifactsToCache(ctx, imageRefDigest, sociIndexDigest, ztocDigests); err != nil {
+	configHashHex := configHash.Hex
+
+	// Try to only read-pull-index-write once at a time to prevent hammering
+	// the containter registry with a ton of parallel pull requests, and save
+	// apps a bunch of parallel work.
+	workKey := fmt.Sprintf("soci-artifact-store-image-%s", configHashHex)
+	respBytes, err := s.deduper.Do(ctx, workKey, func() ([]byte, error) {
```

you're ignoring this error

---

Whoops, fixed